File: day14/day14.py
from collections import defaultdict, deque
from math import ceil


def parse_input():
    recipes = {}
    with open("input.txt") as f:
        for line in f:
            input, output = line.rstrip().split(" => ")
            product_qty, product_name = output.split()

            ingredients = []
            for chem in input.split(", "):
                qty, name = chem.split()
                ingredients.append((int(qty), name))

            recipes[product_name] = (int(product_qty), ingredients)

    return recipes


recipes = parse_input()


def needed_ore(fuel_qty=1):
    queue = deque([(fuel_qty, "FUEL")])
    leftover = defaultdict(int)
    total_ore = 0

    while queue:
        needed_qty, chemical = queue.popleft()

        if chemical == "ORE":
            total_ore += needed_qty
            continue

        if leftover[chemical] >= needed_qty:
            leftover[chemical] -= needed_qty
            continue

        needed_qty -= leftover[chemical]
        leftover[chemical] = 0

        recipe_qty, ingredients = recipes[chemical]
        multiplier = ceil(needed_qty / recipe_qty)

        for qty, el in ingredients:
            queue.append((qty * multiplier, el))

        leftover[chemical] += multiplier * recipe_qty - needed_qty

    return total_ore


ore = needed_ore()
print(ore)

MAX_ORE = 10 ** 12

high = 2
while needed_ore(high) < MAX_ORE:
    high *= 2

low = high // 2

while high - low > 1:
    guess = (low + high) // 2
    ore = needed_ore(guess)

    if ore > MAX_ORE:
        high = guess
    else:
        low = guess

print(low)

File: day14/day14.js
import { rawInput } from './input.js';
// import { exampleInput as rawInput } from './input.js';

const reactions = rawInput
    .split('\n')
    .map((line) => line.split(' => ').map((r) => r.split(', ').map((x) => x.split(' '))));

const recipes = {};

// parse the input into {outputType: {numProduced, inputs: [inputType, inputNeeded]}}
// 7 A, 1 B => 1 C becomes
// {C: {producedPerReaction: 1, input: [[A, 7], [B, 1]]}}
reactions.forEach(([inputs, [output]]) => {
    const input = inputs.map(([amount, inputType]) => [inputType, Number(amount)]);
    const [produced, outputType] = output;
    recipes[outputType] = { producedPerReaction: Number(produced), input };
});

// recursively go through the reactions for the amount of ore needed to produce the given fuel
function calculateOre(fuelToProduce = 1) {
    const extraChemicals = {};
    let totalOre = 0;

    function recursiveLookup(type, quantity) {
        // ore is the base case
        if (type === 'ORE') {
            totalOre += quantity;
            return;
        }

        // check for leftover chemicals we can use
        if (extraChemicals[type]) {
            if (extraChemicals[type] >= quantity) {
                extraChemicals[type] -= quantity;
                return;
            }
            quantity -= extraChemicals[type];
            extraChemicals[type] = 0;
        }

        const { producedPerReaction, input } = recipes[type];

        // how many times the reaction needs to be run rounded up
        const numOfReactions = Math.ceil(quantity / producedPerReaction);

        // save any leftover chemicals
        extraChemicals[type] = (extraChemicals[type] ?? 0) + numOfReactions * producedPerReaction - quantity;

        // for each needed input type lookup how much needs to be made with the correct quantity
        input.forEach(([inputType, amountForReaction]) => {
            recursiveLookup(inputType, numOfReactions * amountForReaction);
        });
    }

    recursiveLookup('FUEL', fuelToProduce);
    return totalOre;
}

// binary search for most fuel using 1 trillion ore
function findMaxFuel() {
    const maxOre = 1000000000000;

    let upper = 2;
    while (calculateOre(upper) < maxOre) {
        upper *= 2;
    }

    let lower = Math.floor(upper / 2);
    while (upper - lower > 1) {
        const guess = Math.floor((lower + upper) / 2);

        if (calculateOre(guess) > maxOre) {
            upper = guess;
        } else {
            lower = guess;
        }
    }
    return lower;
}

const partOne = calculateOre();
const partTwo = findMaxFuel();

console.log(partOne);
console.log(partTwo);
document.getElementById('partOne').appendChild(document.createTextNode(partOne));
document.getElementById('partTwo').appendChild(document.createTextNode(partTwo));
